src/Features/TerrainShadows.h

```diff
@@ -48,57 +35,32 @@ struct TerrainOcclusion : public Feature
 	std::unordered_map<std::string, HeightMapMetadata> heightmaps;
 	HeightMapMetadata* cachedHeightmap;
 
-	struct AOGenBuffer
-	{
-		float AoDistance;  // in game unit
-		uint SliceCount;
-		uint SampleCount;
-
-		float3 pos0;
-		float3 pos1;
-		float2 zRange;
-	};
-	std::unique_ptr<Buffer> aoGenBuffer = nullptr;
-
 	struct ShadowUpdateCB
 	{
 		float2 LightPxDir;   // direction on which light descends, from one pixel to next via dda
 		float2 LightDeltaZ;  // per LightUVDir, upper penumbra and lower, should be negative
 		uint StartPxCoord;
 		float2 PxSize;
-
-		float pad;
+		uint pad0[1];
+		float2 PosRange;
+		float2 ZRange;
 	} shadowUpdateCBData;
 	static_assert(sizeof(ShadowUpdateCB) % 16 == 0);
 	std::unique_ptr<ConstantBuffer> shadowUpdateCB = nullptr;
 
 	struct alignas(16) PerFrame
 	{
 		uint EnableTerrainShadow;
```

Comment:
Don't we usually pass `Settings` directly here so we don't need to duplicate the settings?

Reply:
It's not always enabled